[libunwind][AIX] Remove weak declaration "__xlcxx_personality_v0"

xingxue-ibm opened the pull request:
`__xlcxx_personality_v0` is the personality routine in `libc++abi` for the EH of applications generated by the legacy IBM C++ compiler. Since the EH info generated by the legacy compiler does not provide the location of the personality routine, this routine is hard-coded as the handler for legacy EH in the unwinder. The symbol is resolved dynamically using `dlopen()` to avoid a hard dependency of `libunwind` on `libc++abi` for cases such as non-C++ applications. The weak declaration of `__xlcxx_personality_v0` was originally intended to bypass `dlopen()` if the C++ application generated by the legacy compiler is statically linked with the new LLVM C++ compiler. Unfortunately, this causes problems with runtime linking for Clang-compiled code using the unwinder that does not link with `libc++abi`.

On the other hand, the C++ runtime libraries shipped for AIX are actually stripped and statically linking is not supported. So, we can fix the problem by removing the `__xlcxx_personality_v0` weak declaration. Besides, `dlopen()` would work as long as the libc++abi shared library is available.

xingxue-ibm committed:
Remove weak definition "__xlcxx_personality_v0".

xingxue-ibm committed:
Fix formatting.

hubert-reinterpretcast reviewed (commented):
I think a small test case could be added with a `int main(void) {}` C program linked with -`brtl` and `LDR_PRELOAD`/`LDR_PRELOAD64` set to `libunwind.a(libunwind.so.1)`.

hubert-reinterpretcast commented:
@xingxue-ibm, I edited the PR description text. Please check my edits.

xingxue-ibm commented:
> @xingxue-ibm, I edited the PR description text. Please check my edits.
Thanks for the editing, @hubert-reinterpretcast!

xingxue-ibm committed:
Addressed comments:
- fix comments
- add a LIT test case
- will put driver-by fix to remove dlclose() in a separate PR

xingxue-ibm commented on libunwind/src/UnwindCursor.hpp:
```diff
@@ -2126,42 +2125,34 @@ bool UnwindCursor<A, R>::getInfoFromTBTable(pint_t pc, R &registers) {
 
     // Resolve the function pointer to the state table personality if it has
-    // not already.
+    // not already done.
```
Fixed, thanks!

xingxue-ibm commented on libunwind/src/UnwindCursor.hpp:
```diff
@@ -2126,42 +2125,34 @@ bool UnwindCursor<A, R>::getInfoFromTBTable(pint_t pc, R &registers) {
-            _LIBUNWIND_TRACE_UNWINDING("dlsym() failed with errno=%d\n", errno);
-            assert(0 && "dlsym() failed");
-          }
-          dlclose(libHandle);
```
Restored the `dlclose` call. Will do it in a separate PR.

xingxue-ibm commented on libunwind/src/UnwindCursor.hpp:
```diff
@@ -2126,42 +2125,34 @@ bool UnwindCursor<A, R>::getInfoFromTBTable(pint_t pc, R &registers) {
-        // has been resolved at the link time.
-        xlcPersonalityV0 = &__xlcxx_personality_v0;
+        // Resolve __xlcxx_personality_v0 using dlopen().
+        const char libcxxabi[] = "libc++abi.a(libc++abi.so.1)";
```
Changed as suggested, thanks!

xingxue-ibm commented:
> I think a small test case could be added with a `int main(void) {}` C program linked with -`brtl` and `LDR_PRELOAD`/`LDR_PRELOAD64` set to `libunwind.a(libunwind.so.1)`.

Added a test case `aix_runtime_link.pass.cpp` although the test case needs to be in C. The reason for the `.cpp` suffix is the libunwind test configuration does not recognize `.c` as a suffix for test cases. Also, `%{cxx}` is used to compile. However, the test configuration uses `-nostdlib++ -L %{lib} -lunwind -ldl -Wl,-bbigtoc` as link flags so it does not link with `libc++abi` and the test case does fail without the change.

hubert-reinterpretcast reviewed (approved):
LGTM! Thanks!

xingxue-ibm commented:
Since this is an AIX-specific issue and the fix has been approved by AIX compiler and runtime expert @hubert-reinterpretcast (thank you), I plan to land it tomorrow if there are no objections.
